src/json/from_json.cr

```diff
@@ -8,7 +8,7 @@
 # Int32.from_json("1")                # => 1
 # Array(Int32).from_json("[1, 2, 3]") # => [1, 2, 3]
 # ```
-def Object.from_json(string_or_io : String | IO) : Object
+def Object.from_json(string_or_io : String | IO)
```

Comment:
wouldn't `def Object.from_json(string_or_io : String | IO) : self` also work? 

Not that these annotations add a lot of value, but other `from_json` implementations in this file do it too, eg: https://github.com/crystal-lang/crystal/pull/16142/files#diff-e15cd585e688fc2da5c744812edcda737e746e2c63a991aa346a24ad6175b42cR160

Reply:
Interesting, we would need to be able to refer to the target type of the converter 🤔

Reply:
The converter is only supposed to define `#from_json(JSON::PullParser)`. Those specs are testing `#from_json(String)` instead, which is never supposed to work in the first place because `Enum::ValueConverter` itself does not include `JSON::Serializable`. So the specs are wrong and should be rewritten.

Reply:
The other converters in the standard library are defined in `serializable_spec.cr` instead, so the specs for `Enum::ValueConverter` would have looked like this:

```crystal
class JSONAttrWithEnumValue(T)
  include JSON::Serializable

  @[JSON::Field(converter: Enum::ValueConverter(T))]
  property value : T
end

JSONAttrWithEnumValue(JSONSpecEnum).from_json(%({"value": 1})).value.should eq(JSONSpecEnum::One)
```

Reply:
@HertzDevil I don't think that would work, unfortunately:

```
@[JSON::Field(converter: Enum::ValueConverter(T))]
                                              ^
Error: undefined constant T
```

https://carc.in/#/r/i8sf

Reply:
I've refactored specs in question and reinstated `self` as return type restriction for `Object.from_json` methods.

Reply:
I'm still not convinced about the usefulness of such a restriction tbh. I can imagine a scenario in which I might want to return a proxy object instead of `self`, for instance, and this restriction will prevent me from doing that.